source/common/stats/recent_lookups.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <atomic>
+#include <deque>
+#include <functional>
+#include <utility>
+
+#include "envoy/common/time.h"
+
+namespace Envoy {
+namespace Stats {
+
+// Remembers the last 'Capacity' items passed to lookup().
```

Comment:
I have some high level questions before doing a deep-dive review. 
1) Does time really matter? Would it be better to actually keep a map of name to count with some bounded size with dropping the least counted stats?
2) With the stat set integration, would it be worth it to somehow track the set name and its counts differently since it's a different mutex? Also, how does this relate to the built-in/dynamic stuff in the other PR.
3) I really think we need some type of stat that tracks overall stat name lookup in the slow path, otherwise how would someone know that something might be wrong? Could we have some type of global lookup counter that could be monitored?

Reply:
1. I think time matters in that I don't want to see any of the initial burst of lookups which have to occur on startup, not on the hot path.

2. I am tracking them separately; everyone has separate RecentLookups deques. But they could be displayed separately rather than interleaved together in the admin console.

3. I don't think we know exactly what path we are in. It could be a one-time registration of some interesting new names in response to an xDS event, or it could be in response to a request. From inside the stats APIs I don't think I know (do I?). Anyway the RecentLookups structure does count lookups and we could emit those in admin /stats. I think I was planning to do that and didn't get to it. I can add it. Are you thinking we should also track nanoseconds spent and report that too? That might be interesting.

Reply:
Oh but I guess the absolute time doesn't matter; I could solve that with a sequence number instead, and thus avoid having to pass in the TimeSource and make system calls. But if we're going to measure the real-time consumed we'll be collecting a timestamp anyway.

Reply:
> Oh but I guess the absolute time doesn't matter; I could solve that with a sequence number instead, and thus avoid having to pass in the TimeSource and make system calls. But if we're going to measure the real-time consumed we'll be collecting a timestamp anyway.

TBH I'm not sure I fully grok what is going on here which may be an indication we should try to simplify. 

I think as a user what I really want to know is whether some stat name keeps getting looked up over and over again. I don't really care about single lookups, and I'm not sure I care about time either. So if we kept a map of name to count of lookups and just output the raw count from max, dropping off at a max table length, wouldn't this give me enough visiblity?

For the stat itself, I was thinking about just a raw counter of lookups under lock, which again we could roughly correlate to a lot of steady state lookups vs. just a small burst at startup?

Reply:
Actually in some ways the pending log change simplifies things. I gave up on making RecentLookups templated and just hold a string, to avoid locking issues during logging. I think I'll go ahead and push that; we can always back it out or decide to make a command-line option.

map<name,count> works functionally but how would you want to evict?  I didn't want to keep all the names in the map because that would erase the memory benefit from this work; I wanted to have bounded memory use.

I also thought of putting in a small lru-cache impl but settled on a deque as a cheaper good-enough approximation of that.

At one point in this PR I did collate the deque and print counts. I could bring that back; I'd just collate while printing. In that case i might increase the deque size from 10 to 100 so there's a more to correlate.

The reason I thought of printing the results with the timestamp is that if the lookups were pretty sparse you could correlate the timestamps in the admin console with the server-start time and if all the lookups happened right after startup or xDS update then you could ignore them.

One thing for sure: I'm not really sure how useful the logging or admin console will be until some traffic goes through. I'm sure we'll want to tweak it once there's some mileage on it.

Reply:
What about a bounded max heap and then just sort by count? All of the single lookups would just drop off?

Reply:
Matt & I discussed on Slack a bit ... possibilities include some kind of LFU (heap + map) or LRU (list + map). I find LRU easier to reason because I don't know how to admit new entries into a pure LFU.

I am concerned about the raw low-level speed of the data structure because this is running under the global lock potentially, and so I added a speed test. I got this about 10ns per lookup overhead for the deque, and about 268ns per lookup for the LRU structures. Note that the LRU is all O(1) complexity but a bunch of pointers & hashes to update on every lookup, so the constant-time overhead is 20x that of the deque.

I added the LRU impl in https://github.com/jmarantz/envoy/blob/remember-recent-lookups-lru/source/common/stats/recent_lookups.h and .cc

I added the speed test in this PR. Nothing here is that well tested yet but that's my first cut.

I'm leaning toward the deque impl in this PR as it reports the same data (items and counts) with the same API, with much less overhead. The drawback of the deque is that some potentially interesting entries will age out. But if the idea is that we should be eliminating all those anyway, it should be fine.

Reply:
@jmarantz we should ship and iterate on the dequeue. So will you drop time then and just have items and counts?